### PR TITLE
[core] flink file clean action

### LINKED ISSUE #1571: [Feature] Introduce file clean action
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Currently, the `FileStoreExpireImpl` delete file quietly, which means if there occurs error, the process won't throw exceptions. It's possible that some files won't be deleted when snapshot expiring, so I propose to introduce an action to clean files unused by current snapshots. 

### Solution

Scan all files and filter out that still used by at least one snapshot, then delete those not used by any snapshot.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by john8628**
### Purpose

Linked issue: close #1571

### CODE REVIEW

**Comment by zhangjun0x01** [comment]
hi ,@john8628 , could you add the test case and document ?

**Comment by zhangjun0x01** [comment]
I think we should not only check the data files, but also check metadata

**Comment by john8628** [comment]
> hi ,@john8628 , could you add the test case and document ?
Thank you for your review. I will incorporate these shortly, once I am certain about the legality of my implementation.

> I think we should not only check the data files, but also check metadata
yean metadata should not deleted;  I will implement checks to ensure they are not deleted and remain intact.

**Comment by zhangjun0x01** [comment]
In addition, it is also very important to add an    `olderThan` ,  because the process of writing data is to first write data files. and then write snapshot. When we have written the data but have not yet submitted the snapshot, if the clean action is running at this time, it will delete the newly written data.

**Comment by john8628** [comment]
> In addition, it is also very important to add an `olderThan` , because the process of writing data is to first write data files. and then write snapshot. When we have written the data but have not yet submitted the snapshot, if the clean action is running at this time, it will delete the newly written data.

Thanks for your raising this important consideration;  You're absolutely right, including an "olderThan" parameter is essential in order to prevent the accidental deletion of newly written data when the clean action is executed during the period between writing data files and submitting snapshot;

**Comment by yuzelin** [comment]
Hi, please run `mvn spotless::apply` in your dev environment to fix Checkstyle violation.

**Comment by john8628** [comment]
> Hi, please run `mvn spotless::apply` in your dev environment to fix Checkstyle violation.

OK , thanks , i will do it ;

**Comment by john8628** [comment]
@yuzelin @zhangjun0x01 @JingsongLi PTAL

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanAction.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                "Scan all files and filter out that still used by at least one snapshot, then delete those not used by any snapshot..");
+        // 1: get all files  of the table path
+        List<String> allFileList = buildAllFileList(basePath.getPath());
+        // 2: get all the used files of the table;exclude the manifest , schema  and snapshot files;
+        List<String> validFileNameList = buildValidFileNameList();
+        // 3: get the diff of  the files
+        allFileList.remove(validFileNameList);
+        LOG.info("orphan files:{}", allFileList.size());
+        // 4:delete the file
+        allFileList.stream()
```
Comment (wg1026688210) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanAction.java:84]: How about adding a thread num opt.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanAction.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+/** File Clean table orphan file action for Flink. */
+public class FileCleanAction extends TableActionBase {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FileCleanAction.class);
+    private static final int MAX_DRIVER_LISTING_DEPTH = 3;
```
Comment (wg1026688210) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanAction.java:46]: Do we need to infer based on the number of partition fields and bucket directories?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanActionFactory.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        System.out.println();
+
+        System.out.println("Syntax:");
+        System.out.println(
+                "  file-clean --warehouse <warehouse-path> --database <database-name> "
+                        + "--table <table-name> ");
+
+        System.out.println("Examples:");
+        System.out.println(
+                "  file-clean --path hdfs:///path/to/warehouse/test_db.db/test_table --where id > (SELECT count(*) FROM employee)");
```
Comment (wg1026688210) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/FileCleanActionFactory.java:60]: what 's the usage of `where` opt

**Review by wg1026688210 — COMMENTED** [review]
Left some minor comments

**Comment by john8628** [comment]
> Left some minor comments

thanks for your reply, good  idea；

**Comment by john8628** [comment]
sorry , i can't continue to work for this PR ,

**Comment by zhangjun0x01** [comment]
> sorry , i can't continue to work for this PR ,

I would like to do it .

**Pull request closed without merging**